src/Synth/ADnote.cpp

```diff
@@ -1452,125 +1689,192 @@ inline void ADnote::ComputeVoiceOscillatorRingModulation(int nvoice)
             }
-            vce.FMoldsmp[k] = fmold;
         }
-    }
-    else {  //Phase or PWM modulation
-        const float normalize = synth.oscilsize_f / 262144.0f;
-        for(int k = 0; k < vce.unison_size; ++k) {
-            float *tw = tmpwave_unison[k];
-            for(int i = 0; i < synth.buffersize; ++i)
-                tw[i] *= normalize;
+        else {  //Phase or PWM modulation
+            const float normalize = synth.oscilsize_f / 262144.0f;
+            for(int k = 0; k < vce.unison_size; ++k) {
+                float *tw = tmpwave_unison[k];
+                for(int i = 0; i < synth.buffersize; ++i)
+                    tw[i] *= normalize;
+            }
         }
     }
 
     //do the modulation
     for(int k = 0; k < vce.unison_size; ++k) {
-        float *smps   = NoteVoicePar[nvoice].OscilSmp;
+        float *smps   = NoteVoicePar[nvoice].OscilSmp.smps;
         float *tw     = tmpwave_unison[k];
         int    poshi  = vce.oscposhi[k];
         int    poslo  = (int)(vce.oscposlo[k] * (1<<24));
         int    freqhi = vce.oscfreqhi[k];
         int    freqlo = (int)(vce.oscfreqlo[k] * (1<<24));
 
         for(int i = 0; i < synth.buffersize; ++i) {
-            int FMmodfreqhi = 0;
-            F2I(tw[i], FMmodfreqhi);
-            float FMmodfreqlo = tw[i]-FMmodfreqhi;//fmod(tw[i] /*+ 0.0000000001f*/, 1.0f);
-            if(FMmodfreqhi < 0)
-                FMmodfreqlo++;
-
-            //carrier
-            int carposhi = poshi + FMmodfreqhi;
-            int carposlo = (int)(poslo + FMmodfreqlo);
-            if (FMmode == FMTYPE::PW_MOD && (k & 1))
-                carposhi += NoteVoicePar[nvoice].phase_offset;
-
-            if(carposlo >= (1<<24)) {
-                carposhi++;
-                carposlo &= 0xffffff;//fmod(carposlo, 1.0f);
+
+            if(FMmode == FMTYPE::WAVE_MOD)
+            {
+                // tw[i] is the original modulator
+                // FMSmpMax makes sure tw[i] is in range [-1,1]
+                // Also, FMSmpMax normalizes this, so the oscillating range is
+                // [-basefuncpar, +basefuncpar] or [1-basefuncpar, -1+basefuncpar]
+                // adding basefuncpar finally leads us to
+                // [0, 2*basefuncpar] or [1 - 2*basefuncpar, 1]
+                // The min()/max() calls are just for safety
+                float par = 0.0f;
+                if(pars.VoicePar[nvoice].PWaveEnvelopeEnabled)
+                    par = std::max(0.0f, std::min(1.0f,
+                        (INTERPOLATE_AMPLITUDE(vce.WAVEoldPar,
+                         vce.WAVEnewPar, i, synth.buffersize)) +
+                         NoteVoicePar[nvoice].basefuncpar));
+                else
+                    par = std::max(0.0f, std::min(1.0f,
+                        (tw[i]/NoteVoicePar[nvoice].FMSmpMax) +
+                        NoteVoicePar[nvoice].basefuncpar));
+                // 127 * 4 + 1 waves = 509 waves (3 waves above are padding)
```

Comment:
ok, so 127 is the number of base wave par values (or whatever the wavetable is transforming over. If it's just that, then I don't understand why there's the *4. I'm speculating the +1 is the parameters as exactly the user has specified them, but wouldn't that be accounted in the initial 127? If this is beyond just the basepar shape values then I'm not sure we have a hard and fast reason to bind ourselves to 127, but I haven't reread through oscilgen yet, which might clarify this.

Reply:
The idea was that the usual 128 values are not smooth enough, so instead of "0, 1, 2, 3...", we use 3 values in between each natural number pair, i.e. "0, 0.25, 0.5, 0.75, 1, 1.25 ...". For 128 natural numbers, that makes 127*4+1. The 3 padding waves are unused, just to allow the comfortable "512".

From generating sounds, I was very happy with 3 values in between each, the sound was fine. Taking less might make the transitions less smooth, taking more will cause more computation time. I also think that 512 might be good for many other parameters (though currently, the code only supports the basewave parameter).

Reply:
ok, that explanation makes sense. Similar to a couple other spots within this PR those 'magic numbers' are notable design decisions so promoting the number to a constant in a header with these details would be handy.